src/python/pants/backend/python/util_rules/interpreter_constraints.py

```diff
@@ -170,80 +176,76 @@ def generate_pex_arg_list(self) -> List[str]:
 
     def includes_python2(self) -> bool:
         """Checks if any of the constraints include Python 2.
 
         This will return True even if the code works with Python 3 too, so long as at least one of
         the constraints works with Python 2.
         """
-        return self._includes_version("2.7")
+        return self._includes_version(2, 7)
 
     def minimum_python_version(self, interpreter_universe: Iterable[str]) -> str | None:
         """Find the lowest major.minor Python version that will work with these constraints.
 
         The constraints may also be compatible with later versions; this is the lowest version that
         still works.
         """
-        for major_minor in sorted(interpreter_universe, key=_major_minor_to_int):
-            if self._includes_version(major_minor):
-                return major_minor
+        for major, minor in sorted(_major_minor_to_int(s) for s in interpreter_universe):
+            if self._includes_version(major, minor):
+                return f"{major}.{minor}"
         return None
 
     def _requires_python3_version_or_newer(
         self, *, allowed_versions: Iterable[str], prior_version: str
     ) -> bool:
+        if not self:
+            return False
         patch_versions = list(reversed(range(0, _EXPECTED_LAST_PATCH_VERSION)))
-        # We only need to look at the prior Python release. For example, consider Python 3.8+
+        # We only look at the prior Python release. For example, consider Python 3.8+
         # looking at 3.7. If using something like `>=3.5`, Py37 will be included.
-        # `==3.6.*,!=3.7.*,==3.8.*` is extremely unlikely, and even that will work correctly as
+        # `==3.6.*,!=3.7.*,==3.8.*` is unlikely, and even that will work correctly as
         # it's an invalid constraint so setuptools returns False always. `['==2.7.*', '==3.8.*']`
         # will fail because not every single constraint is exclusively 3.8.
         prior_versions = [f"{prior_version}.{p}" for p in patch_versions]
         allowed_versions = [
             f"{major_minor}.{p}" for major_minor in allowed_versions for p in patch_versions
         ]
-        for req in self:
+
+        def valid_constraint(constraint: Requirement) -> bool:
             if any(
-                req.specifier.contains(prior) for prior in prior_versions  # type: ignore[attr-defined]
+                constraint.specifier.contains(prior) for prior in prior_versions  # type: ignore[attr-defined]
             ):
                 return False
             if not any(
-                req.specifier.contains(allowed) for allowed in allowed_versions  # type: ignore[attr-defined]
+                constraint.specifier.contains(allowed) for allowed in allowed_versions  # type: ignore[attr-defined]
             ):
                 return False
-        return True
+            return True
+
+        return all(valid_constraint(c) for c in self)
```

Comment:
I tried using `self._includes_version()` here, until realizing that _all_ constraints must be valid, not only 1. This change is meant to make that more clear.